README.md
# express-session-sqlite

[![npm version](https://badge.fury.io/js/express-session-sqlite.svg)](https://badge.fury.io/js/express-session-sqlite) ![built with typescript](https://camo.githubusercontent.com/92e9f7b1209bab9e3e9cd8cdf62f072a624da461/68747470733a2f2f666c61742e62616467656e2e6e65742f62616467652f4275696c74253230576974682f547970655363726970742f626c7565) [![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)
 [![CircleCI](https://circleci.com/gh/theogravity/express-session-sqlite/tree/master.svg?style=svg)](https://circleci.com/gh/theogravity/express-session-sqlite/tree/master)

A session store for `express-session` using SQLite.

Fully unit tested. PRs welcomed.

<!-- TOC -->

- [Install](#install)
- [Usage](#usage)
- [Debugging](#debugging)

<!-- TOC END -->

## Install

`$ npm i express-session-sqlite sqlite3 --save`

## Usage

```typescript
import * as sqlite3 from 'sqlite3'
import * as express from 'express'
import * as session from 'express-session'
import sqliteStoreFactory from 'express-session-sqlite'

const SqliteStore = sqliteStoreFactory(session)
const app = express()

app.use(session({
    store: new SqliteStore({
      // Database library to use. Any library is fine as long as the API is compatible
      // with sqlite3, such as sqlite3-offline
      driver: sqlite3.Database,
      // for in-memory database
      // path: ':memory:'
      path: '/tmp/sqlite.db',
      // Session TTL in milliseconds
      ttl: 1234,
      // (optional) Session id prefix. Default is no prefix.
      prefix: 'sess:',
      // (optional) Adjusts the cleanup timer in milliseconds for deleting expired session rows.
      // Default is 5 minutes.
      cleanupInterval: 300000
    }),
    //... don't forget other expres-session options you might need
}))
```

# Debugging

This module uses `debug` under the name `express-session-sqlite`. When starting up your app, do the following:

`$ DEBUG=express-session-sqlite node app.js`


CHANGELOG.md
## 2.1.1 - Fri May 27 2022 11:43:11

**Contributor:** [Theo Gravity](https://github.com/Theo Gravity)

- Update dependencies, remove deprecated inherits (#16)

This is a slight optimization update.

## 2.0.10 - Sun May 09 2021 19:58:39

**Contributor:** [dependabot[bot]](https://github.com/dependabot[bot])

- Bump node-notifier from 8.0.0 to 8.0.2 (#8)

Bumps [node-notifier](https://github.com/mikaelbr/node-notifier) from 8.0.0 to 8.0.2.
- [Release notes](https://github.com/mikaelbr/node-notifier/releases)
- [Changelog](https://github.com/mikaelbr/node-notifier/blob/master/CHANGELOG.md)
- [Commits](https://github.com/mikaelbr/node-notifier/compare/v8.0.0...v8.0.2)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Theo Gravity <theo@suteki.nu>

## 2.0.9 - Sat May 08 2021 20:41:47

**Contributor:** [dependabot[bot]](https://github.com/dependabot[bot])

- Bump handlebars from 4.7.6 to 4.7.7 (#7)

## 2.0.8 - Sat Dec 12 2020 20:56:27

**Contributor:** [dependabot[bot]](https://github.com/dependabot[bot])

- Bump ini from 1.3.5 to 1.3.8 (#3)

Bumps [ini](https://github.com/isaacs/ini) from 1.3.5 to 1.3.8.
- [Release notes](https://github.com/isaacs/ini/releases)
- [Commits](https://github.com/isaacs/ini/compare/v1.3.5...v1.3.8)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

## 2.0.7 - Thu Sep 24 2020 22:54:55

**Contributor:** [Theo Gravity](https://github.com/Theo Gravity)

- Update packages

## 2.0.6 - Sun Jul 19 2020 23:20:37

**Contributor:** [dependabot[bot]](https://github.com/dependabot[bot])

- Bump lodash from 4.17.15 to 4.17.19 (#1)

Bumps [lodash](https://github.com/lodash/lodash) from 4.17.15 to 4.17.19.
- [Release notes](https://github.com/lodash/lodash/releases)
- [Commits](https://github.com/lodash/lodash/compare/4.17.15...4.17.19)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

## 2.0.5 - Tue Apr 07 2020 04:39:13

**Contributor:** [Theo Gravity](https://github.com/Theo Gravity)

- Update readme

## 2.0.4 - Tue Apr 07 2020 04:35:00

**Contributor:** [Theo Gravity](https://github.com/Theo Gravity)

- Update packages to latest

# 2.0.3 - Mon Mar 23 2020 02:22:57

Update `set` to do an upsert instead of insert.

# 2.0.2 - Mon Mar 23 2020 01:58:05

- Add debug logging

# 2.0.1 - Sun Mar 22 2020 23:55:58

- Remove `sqlite3` as dependency.
- Add new `driver` property to specify the sqlite library to use.

# 1.0.5 - Sun Mar 22 2020 20:15:21

Updated `sqlite` to `4.0.0-beta.6`

# 1.0.4 - Sun Mar 22 2020 11:59:50

Updated `sqlite` to `4.0.0-beta.4`

(I wrote `sqlite@4.0.0-beta.4`, it should be ok.)

# 1.0.3 - Sat Mar 21 2020 07:29:58

Update readme.

# 1.0.2 - Sat Mar 21 2020 06:54:47

First version.


src/__tests__/index.test.ts
/* eslint-env jest */

describe('index', () => {
  it('includes the lib', () => {
    const store = require('../index')
    expect(store).toBeDefined()
  })
})


src/__tests__/SqliteStore.test.ts
/* eslint-env jest */
/* eslint handle-callback-err: "warn" */

import * as sqlite3 from 'sqlite3'
import * as session from 'express-session'
import SqliteStoreFactory from '../SqliteStore'

let store

const SID = 'test-1234'

const SAMPLE_DATA = {
  cookie: {
    originalMaxAge: 12345,
    httpOnly: false,
    expires: ('2020-03-21T03:19:30.868Z' as unknown) as Date,
    path: '/',
    maxAge: 50000
  }
}

const SqliteStore = SqliteStoreFactory(session)

beforeEach(done => {
  // @ts-ignore
  store = new SqliteStore({
    driver: sqlite3.Database,
    path: ':memory:',
    ttl: 10000,
    prefix: 'sess:'
  })

  store.set(SID, SAMPLE_DATA, err => {
    done()
  })
})

afterEach(() => {
  clearInterval(store.cleanupTimer)
})

describe('SqliteStore', () => {
  it('should get a session', done => {
    store.get(SID, (err, data) => {
      expect(data).toMatchObject(SAMPLE_DATA)
      done()
    })
  })

  it('should destroy a session', done => {
    store.destroy(SID, () => {
      store.get(SID, data => {
        expect(data).toBe(null)
        done()
      })
    })
  })

  it('should get all sessions', done => {
    store.set('test2', SAMPLE_DATA, () => {
      store.all((err, data) => {
        expect(data.length).toBe(2)
        expect(data[0]).toMatchObject(SAMPLE_DATA)
        done()
      })
    })
  })

  it('should get the session count', done => {
    store.set('test2', SAMPLE_DATA, () => {
      store.length((err, data) => {
        expect(data).toBe(2)
        done()
      })
    })
  })

  it('should clear all sessions', done => {
    store.clear(() => {
      store.length((err, data) => {
        expect(data).toBe(0)
        done()
      })
    })
  })

  it('should touch a session', done => {
    const testData = {
      cookie: {
        originalMaxAge: 4567,
        httpOnly: true,
        expires: ('2020-03-22T03:19:30.868Z' as unknown) as Date,
        path: '/',
        maxAge: 60000
      }
    }

    store.touch(SID, testData, () => {
      store.get(SID, (err, data) => {
        expect(data).toMatchObject(testData)
        done()
      })
    })
  })

  it('should handle expired sessions', done => {
    const s = new SqliteStore({
      driver: sqlite3.Database,
      path: ':memory:',
      ttl: 0,
      prefix: 'sess:'
    })

    clearInterval(s.cleanupTimer)

    s.set(SID, SAMPLE_DATA, () => {
      s.get(SID, data => {
        expect(data).toBe(null)

        s.length((err, count) => {
          expect(count).toBe(0)
          done()
        })
      })
    })
  })
})


src/__tests__/SqliteStoreBase.test.ts
/* eslint-env jest */

import * as sqlite3 from 'sqlite3'
import { SqliteStoreBase } from '../SqliteStoreBase'

let store

const SID = 'test-1234'

const SAMPLE_DATA = {
  cookie: {
    originalMaxAge: 12345,
    httpOnly: false,
    expires: ('2020-03-21T03:19:30.868Z' as unknown) as Date,
    path: '/',
    maxAge: 50000
  }
}

beforeEach(async () => {
  store = new SqliteStoreBase({
    path: ':memory:',
    ttl: 10000,
    prefix: 'sess:',
    driver: sqlite3.Database
  })

  await store.set(SID, SAMPLE_DATA)
})

describe('SqliteStoreBase', () => {
  it('should get a session', async () => {
    const data = await store.get(SID)

    expect(data).toMatchObject(SAMPLE_DATA)
  })

  it('should destroy a session', async () => {
    await store.destroy(SID)
    const data = await store.get(SID)
    expect(data).toBe(null)
  })

  it('should get all sessions', async () => {
    await store.set('test2', SAMPLE_DATA)
    const data = await store.all()
    expect(data.length).toBe(2)
    expect(data[0]).toMatchObject(SAMPLE_DATA)
  })

  it('should get the session count', async () => {
    await store.set('test2', SAMPLE_DATA)
    const data = await store.length()
    expect(data).toBe(2)
  })

  it('should clear all sessions', async () => {
    await store.clear()
    const data = await store.length()
    expect(data).toBe(0)
  })

  it('should touch a session', async () => {
    const testData = {
      cookie: {
        originalMaxAge: 4567,
        httpOnly: true,
        expires: ('2020-03-22T03:19:30.868Z' as unknown) as Date,
        path: '/',
        maxAge: 60000
      }
    }

    await store.touch(SID, testData)
    const data = await store.get(SID)

    expect(data).toMatchObject(testData)
  })

  it('should handle expired sessions', async () => {
    const s = new SqliteStoreBase({
      path: ':memory:',
      ttl: 0,
      prefix: 'sess:',
      driver: sqlite3.Database
    })

    await s.set(SID, SAMPLE_DATA)

    expect(await s.get(SID)).toBe(null)
    expect(await s.length()).toBe(0)
    expect(await s.removeExpiredSessions())
  })
})


src/SqliteStore.ts
import session from 'express-session'
import {
  SqliteStoreBase,
  AllSessionsResult,
  SqliteStoreParams
} from './SqliteStoreBase'
import Timeout = NodeJS.Timeout

// This had to be done because of the dynamic
// inheritance from session.Store
// I prefer to not include express-session as a
// dependency of this package, so I have to
// do this hack to tell typescript the proper typings
interface ISqliteStore {
  new (config: SqliteStoreParams)
  get: (
    sid: string,
    callback: (err: any, session?: session.SessionData | null) => void
  ) => void
  set: (
    sid: string,
    session: session.SessionData,
    callback?: (err?: any) => void
  ) => void
  destroy: (sid: string, callback?: (err?: any) => void) => void
  all: (
    callback: (
      err: any,
      obj?: { [sid: string]: session.SessionData } | null
    ) => void
  ) => void
  length: (callback: (err: any, length?: number | null) => void) => void
  clear: (callback?: (err?: any) => void) => void
  touch: (
    sid: string,
    session: session.SessionData,
    callback?: (err?: any) => void
  ) => void
}

export default function sqliteStoreFactory (session): ISqliteStore {
  const Store = session.Store

  class SqliteStore extends Store implements ISqliteStore {
    sqliteStore: SqliteStoreBase
    cleanupTimer: Timeout | number

    constructor (config: SqliteStoreParams) {
      super(config)

      this.sqliteStore = new SqliteStoreBase(config)
      this.cleanupTimer = setInterval(async () => {
        try {
          await this.sqliteStore.removeExpiredSessions()
        } catch (e) {
          // ignore
        }
        // 5 mins
      }, config.cleanupInterval || 300000)
    }

    get (
      sid: string,
      callback: (err: any, session?: session.SessionData | null) => void
    ) {
      this.sqliteStore
        .get(sid)
        .then(data => {
          callback(null, data)
        })
        .catch(callback)
    }

    set (
      sid: string,
      session: session.SessionData,
      callback?: (err?: any) => void
    ) {
      this.sqliteStore
        .set(sid, session)
        .then(callback)
        .catch(callback)
    }

    destroy (sid: string, callback?: (err?: any) => void) {
      this.sqliteStore
        .destroy(sid)
        .then(callback)
        .catch(callback)
    }

    // @ts-ignore
    all (callback: (err: any, obj?: AllSessionsResult) => void) {
      this.sqliteStore
        .all()
        .then(data => {
          callback(null, data)
        })
        .catch(callback)
    }

    length (callback: (err: any, length?: number | null) => void) {
      this.sqliteStore
        .length()
        .then(length => {
          callback(null, length)
        })
        .catch(callback)
    }

    clear (callback?: (err?: any) => void) {
      this.sqliteStore
        .clear()
        .then(callback)
        .catch(callback)
    }

    touch (
      sid: string,
      session: session.SessionData,
      callback?: (err?: any) => void
    ) {
      this.sqliteStore
        .touch(sid, session)
        .then(callback)
        .catch(callback)
    }
  }

  // @ts-ignore
  return SqliteStore
}


src/SqliteStoreBase.ts
import session from 'express-session'
import { open, Database } from 'sqlite'
import { join } from 'path'

const SQL = require('sql-template-strings')
const debug = require('debug')('express-session-sqlite')

export interface SqliteStoreParams {
  /**
   * The sqlite3 Database driver to use
   */
  driver: any

  /**
   * The path to the sqlite database.
   * For an in-memory database, specify ':memory:'.
   */
  path: string

  /**
   * Session TTL in milliseconds
   */
  ttl: number

  /**
   * Session id prefix. Default is no prefix
   */
  prefix?: string

  /**
   * Triggers a timer in milliseconds to run a cleanup on expired session rows.
   * Default is 5 minutes.
   */
  cleanupInterval?: number
}

export type AllSessionsResult =
  | session.SessionData[]
  | { [sid: string]: session.SessionData }
  | null

export class SqliteStoreBase {
  hasInit: boolean
  config: SqliteStoreParams
  db: Database
  prefix: string

  constructor (config: SqliteStoreParams) {
    this.hasInit = false
    this.config = config
    this.prefix = config.prefix || ''
    this.db = null
  }

  getSid (sid: string) {
    return this.prefix + sid
  }

  async init () {
    if (!this.db) {
      if (!this.config.driver) {
        throw new Error('express-session-sqlite: driver not defined')
      }

      if (!this.config.path) {
        throw new Error('express-session-sqlite: path not defined')
      }

      debug('Opening sqlite database')

      this.db = await open({
        filename: this.config.path,
        driver: this.config.driver
      })

      debug('Running migrations for the session table')

      await this.db.migrate({
        migrationsPath: join(__dirname, 'migrations')
      })

      this.hasInit = true
    }
  }

  async get (sid: string): Promise<session.SessionData | null> {
    debug(`Getting session: ${sid}`)

    await this.init()
    const time = new Date().getTime()
    const resp = await this.db.get(
      SQL`SELECT * FROM sessions WHERE sid = ${this.getSid(
        sid
      )} AND ${time} < expires`
    )
    if (!resp) {
      return null
    }

    debug(`Session found: ${sid}`)

    return JSON.parse(resp.data)
  }

  async set (sid: string, session: session.SessionData): Promise<void> {
    await this.init()

    const serialized = JSON.stringify(session)
    const msTtl = new Date().getTime() + this.config.ttl

    debug(`Setting session: ${sid}`)

    const stmt = await this.db.prepare(
      'REPLACE INTO sessions (sid, data, expires) VALUES (:sid, :data, :expires)',
      {
        ':sid': this.getSid(sid),
        ':data': serialized,
        ':expires': msTtl
      }
    )

    await stmt.run()
  }

  async destroy (sid: string): Promise<void> {
    await this.init()

    debug(`Destroying session: ${sid}`)

    await this.db.run(SQL`DELETE FROM sessions WHERE sid = ${this.getSid(sid)}`)
  }

  async all (): Promise<AllSessionsResult> {
    await this.init()

    const time = new Date().getTime()

    debug(`Fetching all sessions`)

    const data = await this.db.all(
      `SELECT * FROM sessions WHERE ${time} < expires`
    )

    return data.map(raw => {
      return JSON.parse(raw.data)
    })
  }

  async length (): Promise<number | null> {
    await this.init()

    const time = new Date().getTime()

    debug(`Getting session counts`)

    const data = await this.db.get(
      `SELECT count(*) as total FROM sessions WHERE ${time} < expires`
    )
    return data.total
  }

  async clear (): Promise<void> {
    await this.init()

    debug(`Clearing all sessions`)

    await this.db.run(`DELETE FROM sessions`)
  }

  async touch (sid: string, session: session.SessionData): Promise<void> {
    debug(`Refreshing session: ${sid}`)

    await this.init()
    await this.set(sid, session)

    debug(`Refresh session complete: ${sid}`)
  }

  /**
   * Have to manually call this to remove stale entries
   */
  async removeExpiredSessions () {
    await this.init()
    const time = new Date().getTime()

    debug(`Removing expired sessions`)

    await this.db.run(`DELETE FROM sessions WHERE ${time} > expires`)
  }
}


sample-app.ts
import * as express from 'express'
import * as sqlite3 from 'sqlite3'
import * as session from 'express-session'
import sqliteStoreFactory from './build'

const SqliteStore = sqliteStoreFactory(session)
const app = express()

app.use(session({
  cookie: { maxAge: 86400000 },
  secret: 'secret-for-encoding-cookie',
  store: new SqliteStore({
    // for in-memory database
    // path: ':memory:'
    path: '/tmp/sqlite.db',
    // Session TTL in milliseconds
    ttl: 1234,
    // (optional) Session id prefix. Default is no prefix.
    prefix: 'sess:',
    // (optional) Triggers a timer in milliseconds to run a cleanup on expired session rows.
    // Default is 5 minutes.
    cleanupInterval: 300000,
    driver: sqlite3.Database
  }),
  resave: false,
  saveUninitialized: true
}))

app.get('/', (req, res) => {
  req.session.blah = 'blah'
  req.session.save(() => {
    res.send('ok')
  })
})

app.listen(3000, () => console.log(`Example app started!`))
